File: Graphs/Q3. Max Area of Island/Solution.java
class Solution {
    public int dfs(int row, int col, int grid[][], boolean vis[][]){
        vis[row][col] = true;

        int m = grid.length;
        int n = grid[0].length;
        int dr[] = {-1, 0, 1, 0};
        int dc[] = {0, 1, 0, -1};
        int area = 1;
        for(int i=0; i<4; i++){
            int nrow = row+dr[i];
            int ncol = col+dc[i];
            if(nrow>=0 && ncol>=0 && nrow<m && ncol<n && grid[nrow][ncol]==1 && !vis[nrow][ncol]){
                area += dfs(nrow, ncol, grid, vis);
            }
        }
        return area;
    }

    public int maxAreaOfIsland(int[][] grid) {
        int max = 0;
        int m = grid.length;
        int n = grid[0].length;
        boolean vis[][] = new boolean[m][n];

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]==1 && !vis[i][j]){
                    int area = dfs(i, j, grid, vis);
                    max = Math.max(max, area);
                }
            }
        }

        return max;
    }
}

File: Graphs/Q3. Max Area of Island/solution.cpp
class Solution {
public:
    int dfs(int row, int col, vector<vector<int>>& grid,vector<vector<bool>>&vis ){
        int m=grid.size();
        int n=grid[0].size();

        if(row<0 || col<0 || row>=m || col>=n || vis[row][col]||grid[row][col]!=1){
            return 0;
        }

        vis[row][col]=true;

        int left=dfs(row,col-1,grid,vis);
        int right=dfs(row,col+1,grid,vis);
        int up=dfs(row-1,col,grid,vis);
        int down=dfs(row+1,col,grid,vis);

        return 1+ left+right+up+down;

    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ans=0;
        int m=grid.size();
        int n=grid[0].size();
        vector<vector<bool>>vis(m,vector<bool>(n,false));
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(!vis[i][j] && grid[i][j]==1){
                    int size=dfs(i,j,grid,vis);
                    ans=max(ans,size);
                }

            }
        }
        return ans;
        
    }
};
